feat(traverse): add `is_static` method to `TraverseCtx`, which is moves out from SymbolTable

Dunqing opened the pull request:
The [scope.isStatic](https://github.com/babel/babel/blob/419644f27c5c59deb19e71aaabd417a3bc5483ca/packages/babel-traverse/src/scope/index.ts#L557) method we port it from Babel. This method is useful for `transformer`. So we should move it out from `SymbolTable`, and it would be better to put it to `TraverseCtx`.

Dunqing commented:
* **#5067** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5067?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented:
As @overlookmotel  said in https://github.com/oxc-project/oxc/issues/5049#issuecomment-2304138762, this method also doesn't related to `Traverse`. I don't know where we should put it is

overlookmotel commented on crates/oxc_traverse/src/context/scoping.rs:
```diff
@@ -382,6 +382,30 @@ impl TraverseScoping {
+    pub fn is_static(&self, expr: &Expression) -> bool {
+        match expr {
+            Expression::ThisExpression(_) | Expression::Super(_) => true,
+            Expression::Identifier(ident) => ident.reference_id.get().is_some_and(|reference_id| {
+                self.symbols.get_reference(reference_id).symbol_id().is_some_and(|symbol_id| {
+                    self.symbols.get_resolved_references(symbol_id).all(|r| !r.is_write())
+                })
+            }),
```
I think it'd be better to `unwrap` on `ident.reference_id.get()`. In the transformer, semantic has been run, so all `IdentifierReference`s should have a `ReferenceId`. Using `is_some_and` suggests that it's optional, when it's not.

Dunqing replied:
Great suggestion

overlookmotel reviewed (approved):
I have been thinking that we should create a separate `Scopes` struct which combines `ScopeTree` and `SymbolTable`. The we can move most of `TraverseCtx`'s scope/symbol-related methods to that struct, and they can all live in `oxc_semantic` crate - and be used within transformer and also any other code which has access to `Semantic`.

https://github.com/oxc-project/backlog/issues/92

But we don't have that yet, so I guess `TraverseCtx` is the best place for this method for now.

Dunqing commented on crates/oxc_traverse/src/context/scoping.rs:
```diff
@@ -382,6 +382,35 @@ impl TraverseScoping {
+    /// Based on Babel's `scope.isStatic` logic.
+    /// <https://github.com/babel/babel/blob/419644f27c5c59deb19e71aaabd417a3bc5483ca/packages/babel-traverse/src/scope/index.ts#L557>
+    ///
+    /// ### Panics
+    /// Never panic because [IdentifierReference] must have a reference_id.
```
Should I add this?

Dunqing committed:
feat(traverse): add `is_static` method to `TraverseCtx`, which is moves out from SymbolTable (#5067)

The [scope.isStatic](https://github.com/babel/babel/blob/419644f27c5c59deb19e71aaabd417a3bc5483ca/packages/babel-traverse/src/scope/index.ts#L557) method we port it from Babel. This method is useful for `transformer`. So we should move it out from `SymbolTable`, and it would be better to put it to `TraverseCtx`.
